refactor: modifications and additional tests for weighted distance

lucasmccabe opened the pull request:
Fixes #5257

Additional tests have been added for weighted graph distance computations.

dschult commented:
Rather than a boolean input using a fixed attribute to hold the weights, we want this to allow input of the attribute name that should be used as a weight, and even allow a weight function.  See the treatment done for weight arguments in the `algorithms/shortest_paths/weighted.py` module.

lucasmccabe commented:
Thanks for the ping @rossbar. In the most recent commit, I modeled the weight arguments in `distance_measures` after those in `shortest_paths/weighted`. Is this closer to what you and @dschult had envisioned?

lucasmccabe commented:
Hello @rossbar - is this something you think we can consider for review, or is it not what you had in mind?

dschult reviewed (commented):
This looks good! Thanks.  I've got a couple of concerns on the handling of extrema_bounding (see below) and the same comments arise for eccentricity so I didn't repeat them. Basically, we don't want to worry about rounding the results. And we probably want to use the generic shortest_path_length so it can handle whether to compute with weights or not. 

Removing the rounded values in the code might impact the tests, but probably not for these simple examples.

dschult commented:
I resolved some conflicts with the main branch (which removed deprecated code). So you will need to pull the changes from your fork to your local machine before your next commit. Hope that's not trouble and if you need help sorting that out let us know.

lucasmccabe commented on networkx/algorithms/distance_measures.py:
```diff
@@ -163,7 +104,20 @@ def _extrema_bounding(G, compute="diameter"):
+        dist = nx.single_source_bellman_ford_path_length(
+            G=G, source=current, weight=weight
+        )
+        if isinstance(weight, str):
+            edge_weights = nx.get_edge_attributes(G, "weight").values()
+        else:
+            edge_weights = [weight(v, u, e) for u, e in G_succ[n].items()]
+
+        decimal_places = [str(w)[::-1].find(".") for w in edge_weights if w]
+        if decimal_places:
+            decimal_places = max(decimal_places)
+            for k in dist.keys():
+                dist[k] = round(dist[k], decimal_places)
+
```
> Am I correct in inferring that this code is only needed so we can round off the distances to a certain number of decimal places?

You are correct.

> If so, we should leave this out. 

OK, this makes sense. Removing the precision-handling will cause issues with the existing tests, but I can address this by giving the tests integer weights.

> I like the way you figured out the number of decimals for the incoming weights though. :)

Thanks! :)

> And anyone using floating point weights will have to deal with round-off error on their own. 

Do you think it would make sense to add a note to this effect in the docstring?

dschult replied:
I guess it never hurts to remind people about round-off error. Something short -- and maybe where the `weight` input is described. like: "Weights stored as floating point values can lead to small round-off errors in distances. Use integer weights to avoid this."

For the tests, I think you should leave the tests in place testing the round-off error. For each failing assertion, change it to expect the failure (`assert ... != ...` or `assert not ...`). Then right below it add a line that tests that they are close in value. The [pytest utility approx](https://docs.pytest.org/en/latest/reference/reference.html?highlight=approx#pytest.approx) does a sophisticated checking process so you can do `assert expected == pytest.approx(actual)`.  And it checks all the values when you give it lists or dicts.  Anyway, test that round-off is a problem so we can see when that is no longer the case, and also test that the result is "close".

So: `assert nx.eccentricity(self.G, 1) == 15`
would become:
```python
assert nx.eccentricity(self.G, 1) != 15
assert pytest.approx(nx.eccentricity(self.G, 1)) == 15
```

lucasmccabe replied:
> The [pytest utility approx](https://docs.pytest.org/en/latest/reference/reference.html?highlight=approx#pytest.approx) does a sophisticated checking process so you can do assert expected == pytest.approx(actual)

Ooh, I was not aware of this - very cool!

I'll revert to the original float-weighted tests and make this adjustment.

lucasmccabe commented on networkx/algorithms/distance_measures.py:
```diff
@@ -163,7 +104,20 @@ def _extrema_bounding(G, compute="diameter"):
 
         # get distances from/to current node and derive eccentricity
-        dist = dict(nx.single_source_shortest_path_length(G, current))
+        dist = nx.single_source_bellman_ford_path_length(
```
I see what you mean. How about I switch to `shortest_path_length` in `_extrema_bounding` and provide a distance_method arg that defaults to "dijkstra" in `_extrema_bounding`, `diameter`, `periphery`, `radius`, and `center`?

dschult replied:
Upon thinking about this more...  This is the `distance_method` module. The edge weights are distances. So it doesn't make sense for them to be negative. So, we should just use dijkstra for the weighted case.  That is the default for `shortest_path_length`.  So I suggest using `shortest_path_length for all the functions -- and not add an extra arg to all those functions.   

Thoughts?

lucasmccabe replied:
Hm, I'm not sure. I'll defer to your guidance, but since we've given the user the freedom to use any weight function, I feel like we should be able to handle negative weights.

If you think the distance_method arg cluters the api, what if I just check to see if weights are negative in `_extrema_bounding` and `eccentricity`?

dschult replied:
The user can provide an arbitrary function, but it still has to make sense... Just like we let them provide any weight value, but it has to make sense.  And you shouldn't have to check for negative weights. That will be done in the `shortest_path_length`.  But we should probably add another phrase in the weight doc_string description that the weights should be positive (since they are distances).

lucasmccabe replied:
Ok, this makes sense. I've removed the distance_method argument and added a sentence to the relevant docstrings indicating that weights should be positive.

dschult commented:
This looks good.  I checked out how the documentation would look after this PR.  You can do that by clicking on "Show all checks" on the right side of the box that says "All checks have passed".  Then scroll down to the 3rd to last check which is run using "CircleCI" and says "document artifact".  Click on the "Details" button and you are brought to a version of the webpage for documentation as it will be after this PR.  It's a little bit of clicking to find your function: Reference (at the top) then algorithms and "distance measures".  And then you can look at each function and how it renders wit its doc_string.

All your doc_strings look good.

But I do see that "resistance_distance" has some unformatted notes at the bottom.  Would you be willing to change the formatting near lines 672.  After the "Overview discussion:" we need a blank line and then indent the two bullet lines below it.  Similarly, after the heading "Additional Details:", only those also need a star in front to make a bullet in the webpage.  If you have had enough, just say so and we'll stop here and merge this. The other change can be done in another PR.

Thanks very much!  Job well done! :}

lucasmccabe commented:
> This looks good.

Great! Thanks for the review, @dschult.

> But I do see that "resistance_distance" has some unformatted notes at the bottom.

I see what you mean. In the most recent commit, I address this and make a similar adjustment in `_extrema_bounding`.

> Some checks were not successful

I'm not sure what to make of the three cancelled checks. The only differences between [072b960](https://github.com/networkx/networkx/pull/5305/commits/072b9603adf143579de1fb79fcb50ec52d0b5a8c) and [d90a41f](https://github.com/networkx/networkx/pull/5305/commits/d90a41f258424039a3f815f67443b7269642e64a) are the docstring updates, and tests pass locally. Perhaps it has something to do with [this](https://www.githubstatus.com/incidents/437q52fb70hg?utm_ts=1658155156)?

dschult commented:
I think the tests just timed out on the hardware they were use. (There are time limits for some donated hardware.) I have restarted the first and it did not cancel this time. Will restart the other two and hope for the best.

lucasmccabe commented:
> I think the tests just timed out on the hardware they were use. (There are time limits for some donated hardware.) I have restarted the first and it did not cancel this time. Will restart the other two and hope for the best.

Ah, I see. Thanks for your help!

dschult reviewed (commented):
This is really good -- Thanks very much.

I only have one comment/question. In a few of the functions, you call with the first input value `G=G` instead of just `G`.  The function signature doesn't provide a keyword argument for the first input. It's just `G`. It reminds me that recent versions of Python allow  function arguments to be keyword-only or position-only. Is there a reason you used this syntax? I don't know the implications.

dschult commented on networkx/algorithms/distance_measures.py:
```diff
@@ -14,7 +14,7 @@
 
 
-def _extrema_bounding(G, compute="diameter"):
+def _extrema_bounding(G, compute="diameter", weight="weight"):
```
Yes -- it should default to `weight=None`.  

The `weighted.py` module does use `weight="weight"`, but we shouldn't use the `weighted.py` functions unless a `weight` argument is provided. That's what `shortest_path-lengths` does for us (which is in the `generic.py` module). It chooses whether to call the weighted functions, or stick with the unweighted functions with the default choice being unweighted (unless a weight value is provided).

lucasmccabe replied:
> That's what shortest_path_lengths does for us (which is in the generic.py module). It chooses whether to call the weighted functions, or stick with the unweighted functions with the default choice being unweighted (unless a weight value is provided).

I'm not sure I agree @dschult @rossbar . We are trying to distinguish three cases:
1. Weights are via a provided weight function
2. Weights are via edge attributes ( weight="weight")
3. Weights are all 1 ("unweighted" case)

When weight="weight", we're not defaulting to the unweighted case, we're using the edge attributes as weights. If we pass weight=None, we are telling `shortest_path_length` to set weights to 1, not use the edge attributes. This is written in the `shortest_path_length` docstring:

> weight : None, string or function, optional (default = None)
        **If None, every edge has weight/distance/cost 1.**
        If a string, use this edge attribute as the edge weight.
        Any edge attribute not present defaults to 1.
        If this is a function, the weight of an edge is the value
        returned by the function. The function must accept exactly
        three positional arguments: the two endpoints of an edge and
        the dictionary of edge attributes for that edge.
        The function must return a number.

Constructively, three paths forward come to mind:
1. Leave  weight="weight" as the default in all `distance_measures`. Consider refactoring `distance_measures` and `generic` at a later time.
2. In all `_extrema_bounding`, do something like:
```
if not weight:
      dist = nx.shortest_path_length(G=G, source=current, weight="weight")
else:
      dist = nx.shortest_path_length(G=G, source=current, weight=weight)
```
3. refactor `generic` as part of this PR.

My view is refactoring `generic` is out of scope, and #2 seems clunky, but I am happy to do it if you think that best. Do you see a better alternative?

dschult replied:
We want the unweighted values when the user does not provide the `weight` argument to the function.
I think all the code is just fine except for the default value of weight, which should be None just as it is for `shortest_path_length`.  Then for each use case the right thing happens:

1) Weight=function: The user specifies a weight function, so `weight` is not None and `shortest_path_length` uses the weight provided weight function -- all is good.
2) Weights=edge attribute name.  For example, weight="weight" or weight="distance" or weight="cost" or any other attribute name.  Again, weight is not None, so `shortest_path_length` uses the edge attribute specified to find the weights.
3) Unweighted case:  This is the old/existing case. The user doesn't specify any `weight` and so `weight=None` is passed into `shortest_path_length` which calls the unweighted path length function.

rossbar replied:
@lucasmccabe consider the following.

Let's say a user has the following code:

```python
>>> import networkx as nx
>>> G = nx.Graph([(0, 1, {"weight": 10})])
>>> nx.diameter(G)
```

In all previous versions of networkx, this would give `1`. However, if we were to make `weight="weight"` the default, then this would give `10` instead; i.e. we'd be silently changing the results of existing user code without warning. This is something that we must avoid, otherwise we'd potentially be introducing bugs (and hard-to-track-down ones, at that) all over existing user code!

I certainly understand the logic of wanting a weighted result by default if the graph is weighted. Even without backwards-compatibility constraints, choosing an explicit default for an edge attribute doesn't mesh well with the generality of the library. Consider another example: let's say someone has a graph where the nodes represent towns and the edges represent roads between them (as I gather, this is a pretty common example in geospatial applications). The edges of this graph might have multiple attributes, say one called `"distance"` that represents the physical distance between nodes along a given edge, and another called `"weight"` which represents the number of travelers / hour that travel along that edge. In this case, if a user is calling one of the `distance_measures` functions on the graph, chances are they actually want the result to be weighted by the `"distance"` edge attr, but we'd be giving them a different (and harder to understand for this example) value *by default*. In this case too, I'd argue it's much better to have an unweighted default and force the user to be intentional about which weight they want.

Anyways, just some high-level pontification. The real blocker for me with going with `weight="weight"` as the default is the backward-incompatibility concern!

lucasmccabe replied:
@dschult  @rossbar  Thank you both for being patient and explaining your reasoning here; I now understand what you mean.

I thought you wanted to make `weight=None` do what `weight="weight"` does, and drop `weight="weight"` as an option altogether. My mistake!

dschult commented on networkx/algorithms/tests/test_distance_measures.py:
```diff
@@ -97,6 +97,62 @@ def test_eccentricity_directed_weakly_connected(self):
+        assert nx.diameter(self.G) == 1.6
+
+    def test_radius(self):
+        assert nx.radius(self.G) != 0.9
```
OK... that's my mistake.  I was thinking we should check that we understand what is happening by testing that round-off occurs. But I agree that this is not the answer we care about. 

Should we make all the comparisons use the `pytest.approx` idiom? or only the ones where round-off messes up the comparison?

rossbar replied:
> Should we make all the comparisons use the pytest.approx idiom? or only the ones where round-off messes up the comparison?

It's probably safest to use `pytest.approx` everywhere for testing "equality" with floating point values, but it's also not super critical. I doubt it will cause failures except *maybe* on systems that use 32-bit floats.

dschult replied:
Would it be useful to know (and test for) cases where round-off makes a difference. In these examples, 2 have round-off and the rest don't. Do we care if that changes? probably not. :}

dschult commented on networkx/algorithms/tests/test_distance_measures.py:
```diff
@@ -97,6 +97,62 @@ def test_eccentricity_directed_weakly_connected(self):
+        e = nx.eccentricity(self.G, v=1)
+        assert e == 1.5
+
+        # This behavior changed in version 1.8 (ticket #739)
+        e = nx.eccentricity(self.G, v=[1, 1])
+        assert e[1] == 1.5
+        e = nx.eccentricity(self.G, v=[1, 2])
+        assert e[1] == 1.5
```
Interesting... these tests seem to be testing the distinction between returning a value and returning a dict. When the input is an iterable/iterator of nodes, the output should be a dict -- even when there is only one value computed. When the input is a node, then the output should be a value.

The blame on the code (and the commented out code at the top of the function) indicate that this treatment of the input `v` was indeed intended to handle input `v` that are lists. This behavior goes all the way back to the hg(pre-git) repository. So, we built in handling of list inputs without advertising it. Of course, the original docs didn't say anything about the input or output parameters. That was added later (still in the hg days) and didn't discuss inputs other than a node or None, even though the code handled it.

So, I think it's fair to say that the behavior has been enshrined -- which does not mean it is the right choice for API. What would be the right choice?

lucasmccabe replied:
Re: the question is whether to support `v` being a list - I think the answer should be "yes," but I will defer to @dschult  and @rossbar. I could just add to the docstring to indicate that `v` can be a list, as well.

That being said, I do feel like this API decision may be out of scope for this ticket. As it stands, `v` can be a list, so I test this, just as it is tested in the unweighted case.

lucasmccabe commented:
@dschult @rossbar Thank you for your review and guidance. The adjustments recommended above are included in the most recent commit, and tests are implemented for all three `weight` options (None, "weight", function).

dschult commented:
Ack -- looks like you merged with the v2.8 branch instead of the main branch. 
I think we can fix that.  I'll try later today.

The changing of the default weight looks good. 
Can we have a test to make sure it works even if the name of the weight attribute is "cost' (or anything else that is not "weight")?  Go ahead and change and push when you've got it. I'll fix the v2.8 branch stuff after that. (or before it if I get to it before you get to it)

lucasmccabe commented:
> make sure it works even if the name of the weight attribute is "cost"

Good idea - I've added two more edge attributes in the tests: "cost" (which has the same value as "weight") and "high_cost" (which is 10x "cost"). I test that using weight="cost" produces the same results as weight="weight," then test that I get the appropriate results when using weight="high_cost" (i.e. distances like `radius` should be different, but vertex groups like `center` should be the same).

> looks like you merged with the v2.8 branch

My mistake. I reverted the commit that merged with v2.8 and made a separate commit merging with main.

dschult reviewed (approved):
This looks good to me!
Thanks @lucasmccabe for sticking with all our requests for changes.

rossbar reviewed (approved):
Thanks @lucasmccabe ! I took the liberty of making a few final cosmetic changes, but everything else LGTM

lucasmccabe commented:
Great! Thanks for your help @dschult @rossbar
